fem/coefficient.cpp

```diff
@@ -758,4 +758,95 @@ double ComputeGlobalLpNorm(double p, VectorCoefficient &coeff, ParMesh &pmesh,
 }
 #endif
 
+VectorQuadratureFunctionCoefficient::VectorQuadratureFunctionCoefficient(
+   QuadratureFunction *qf)
+   : VectorCoefficient(qf->GetVDim()), QuadF(qf), index(0),
+     length(qf->GetVDim()) {}
+
+void VectorQuadratureFunctionCoefficient::SetQuadratureFunction(
+   QuadratureFunction *qf)
+{
+   index = 0;
+   length = qf->GetVDim();
+   vdim = length;
+   QuadF = qf;
+}
+
+void VectorQuadratureFunctionCoefficient::SetLength(int _length)
+{
+   MFEM_VERIFY(_length > 0, "Length must be > 0");
+
+   int diff = QuadF->GetVDim() - index;
+   MFEM_VERIFY(_length <= diff,
+               "Length must be <= (QuadratureFunction length - index)");
+
+   length = _length;
+   vdim = length;
+}
+
+void VectorQuadratureFunctionCoefficient::SetIndex(int _index)
+{
+   MFEM_VERIFY(_index >= 0, "Index must be >= 0");
+   MFEM_VERIFY(_index < QuadF->GetVDim(),
+               "Index must be < the QuadratureFunction length");
+   index = _index;
+   // check to see if length needs to be modified
+   int diff = QuadF->GetVDim() - index;
+   if (length > diff)
+   {
+      length = diff;
+      vdim = length;
+   }
+}
+
+void VectorQuadratureFunctionCoefficient::Eval(Vector &V,
```

Comment:
At exit, `V` is pointing to the `data` of the `QuadF`. This is somewhat subtle, and the used might not expect that changing the values of `V` will update the `QuadratureFunction`.

Reply:
Nice catch it shouldn't be a problem creating a copy of the data and handing that back. I'll work on a fix for that.

Reply:
One way to catch issues like that is to make `QuadF` a `const QuadratureFunction *`. It makes sense that the coefficient will only need const access to the `QuadratureFunction`.